proof/invariant-abstract/IpcDet_AI.thy

```diff
@@ -87,113 +87,109 @@ lemma sfi_tcb_at [wp]:
     \<lbrace>\<lambda>_. tcb_at t\<rbrace>"
   by (wpsimp simp: send_fault_ipc_def)
 
-lemma sort_queue_valid_ep_helper:
-  "distinct list \<Longrightarrow> (a, b) \<in> set (zip list' list) \<Longrightarrow> (a', b) \<in> set (zip list' list) \<Longrightarrow> a = a'"
-  apply (induct list arbitrary: list')
-   apply clarsimp
-  apply (clarsimp simp: zip_Cons)
-  apply (erule disjE, fastforce dest!: in_set_zipE)
-  apply (erule disjE, fastforce dest!: in_set_zipE, clarsimp)
+lemma tcb_ep_find_index_wp:
+  "\<lbrace>\<lambda>s. (\<forall>i j. 0 \<le> i \<and> i \<le> Suc sz \<longrightarrow>
+               (\<forall>tcb tcba. ko_at (TCB tcb) tptr s \<and> ko_at (TCB tcba) (queue ! j) s \<longrightarrow>
+                           (Suc j = i \<longrightarrow> tcb_priority tcba \<ge> tcb_priority tcb) \<longrightarrow>
+                           (i < j \<and> j \<le> sz \<longrightarrow> tcb_priority tcba < tcb_priority tcb) \<longrightarrow> Q i s))\<rbrace>
+   tcb_ep_find_index tptr queue sz \<lbrace>Q\<rbrace>"
+  by (induct sz) (wp thread_get_wp' | simp add: tcb_ep_find_index.simps obj_at_def le_Suc_eq)+
```

Comment:
nitpick: it's nicer to use `wpsimp wp: foo simp: bar` instead of `(wp foo | simp add: bar)+` when possible.

Reply:
I did try that, not possible unfortunately.